urla5230a98e3_Dhana_Krishnasamy_InMemoryJMS_tgz-pJ8-com_dhana_broker_demo_AppJ8/src/com/dhana/broker/impl/AbstractMessageConsumer.java

```diff
@@ -0,0 +1,180 @@
+		AsyncMessageConsumer {
+	static Logger logger = LoggerFactory
+			.getLogger(AbstractMessageConsumer.class);
+	private static Message poison = new PoisonMessage();
+	final LinkedBlockingQueue<Message> queue;
+	protected Destination destination;
+	private MessageDiscardPolicy discardPolicy;
+	private static MessageDiscardPolicy defaultDiscardPolicy = MessageDiscardPolicy.DiscardLatest;
+	private static int defaultQueueSize = 10;
+	private Executor timedReciveExecutor = Executors.newCachedThreadPool();
+	private MessageExpiryUtil expiryUtil;
+	private volatile boolean isClosed = false;
+
+	AbstractMessageConsumer(Destination destination)
+			throws DestinationClosedException {
+		this(destination, defaultQueueSize, defaultDiscardPolicy);
+	}
+
+	AbstractMessageConsumer(Destination destination, int queueSize)
+			throws DestinationClosedException {
+		this(destination, queueSize, defaultDiscardPolicy);
+	}
+
+	AbstractMessageConsumer(Destination destination, int queueSize,
+			MessageDiscardPolicy discardPolicy)
+			throws DestinationClosedException {
+		if (destination == null)
+			throw new AssertionError("Topic cannot be null");
+		this.destination = destination;
+		this.discardPolicy = discardPolicy;
+		queue = new LinkedBlockingQueue<Message>(queueSize);
+		expiryUtil = new MessageExpiryUtil(queue);
+		destination.addSubscriber(this);
+	}
+
+	/** Topic callback **/
+	public void onMessage(Message msg) {
+		if (msg == null)
+			throw new AssertionError("message cannot be null");
+		boolean isSuccessful = queue.offer(msg);
+		if (!isSuccessful) {
+			if (discardPolicy == MessageDiscardPolicy.DiscardOldest) {
+				// Until successful insert
+				while (!queue.offer(msg)) {
+					queue.poll();
+				}
+				expiryUtil.scheduleForExpiry(msg);
+			}
+		} else {
+			expiryUtil.scheduleForExpiry(msg);
+		}
+	}
+
+	/** Client methods **/
+	public void close() {
+		logger.info("Closing MessageConsumer");
+		destination.removeSubscriber(this);
+		queue.offer(poison);
+		queue.clear();
+		expiryUtil.cancel();
+		isClosed = true;
+	}
+
+	public Message receive() throws InterruptedException,
+			DestinationClosedException {
+		if (isClosed)
+			throw new DestinationClosedException("MessageConsumer closed");
+		Message m;
+		// Make sure expired message is not delivered
+		while (true) {
+			m = queue.take();
+			if (m == poison) {
+				logger.info("Close signal received");
+				throw new DestinationClosedException("MessageConsumer closed");
+			}
+			if (!m.hasExpired())
+				return m;
+			else
+				logger.debug("Message Expired:" + m);
+		}
+	}
+
+	public Message receive(int timeOutMilliseconds)
+			throws InterruptedException, DestinationClosedException {
+		if (isClosed)
+			throw new DestinationClosedException("MessageConsumer closed");
+		FutureTask<Message> future = getReceiveFuture();
+		timedReciveExecutor.execute(future);
+		try {
+			return future.get(timeOutMilliseconds, TimeUnit.MILLISECONDS);
+		} catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)